src/storage/inspector.rs

```diff
@@ -0,0 +1,199 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::fmt::{Debug, Display, Formatter, Result as FmtResult};
+use std::sync::atomic::{self, AtomicU64};
+use std::sync::{Arc, RwLock};
+
+use raft::StateRole;
+
+use crate::raftstore::coprocessor::{
+    Coprocessor, CoprocessorHost, ObserverContext, RegionChangeEvent, RegionChangeObserver,
+    RoleObserver,
+};
+use crate::util::collections::HashMap;
+use crate::util::worker::Scheduler;
+use crate::util::HandyRwLock;
+
+//const KEY_BUCKET_SIZE: usize = 1024;
+
+#[derive(Debug, Clone)]
+enum LeaderChangeEvent {
+    Update { region_id: u64, version: u64 },
+    Remove { region_id: u64 },
+}
+
+impl Display for LeaderChangeEvent {
+    fn fmt(&self, fmt: &mut Formatter) -> FmtResult {
+        Debug::fmt(self, fmt)
+    }
+}
+
+#[derive(Clone)]
+struct LeaderChangeObserver {
+    scheduler: Scheduler<LeaderChangeEvent>,
+}
+
+impl LeaderChangeObserver {
+    //    pub fn new(scheduler: Scheduler<LeaderChangeEvent>) -> Self {
+    //        Self { scheduler }
+    //    }
+}
+
+impl Coprocessor for LeaderChangeObserver {}
+
+impl RoleObserver for LeaderChangeObserver {
+    fn on_role_change(&self, ctx: &mut ObserverContext, role: StateRole) {
+        if role == StateRole::Leader {
+            self.scheduler
+                .schedule(LeaderChangeEvent::Update {
+                    region_id: ctx.region().get_id(),
+                    version: ctx.region().get_region_epoch().get_version(),
+                })
+                .unwrap();
+        } else {
+            self.scheduler
+                .schedule(LeaderChangeEvent::Remove {
+                    region_id: ctx.region().get_id(),
+                })
+                .unwrap();
+        }
+    }
+}
+
+impl RegionChangeObserver for LeaderChangeObserver {
+    fn on_region_changed(
+        &self,
+        ctx: &mut ObserverContext,
+        event: RegionChangeEvent,
+        role: StateRole,
+    ) {
+        if role == StateRole::Leader {
+            let event = match event {
+                RegionChangeEvent::Create | RegionChangeEvent::Update => {
+                    LeaderChangeEvent::Update {
+                        region_id: ctx.region().get_id(),
+                        version: ctx.region().get_region_epoch().get_version(),
+                    }
+                }
+                RegionChangeEvent::Destroy => LeaderChangeEvent::Remove {
+                    region_id: ctx.region().get_id(),
+                },
+            };
+            self.scheduler.schedule(event).unwrap();
+        }
+    }
+}
+
+// region_id -> (epoch_version, max_read_ts)
+type TsMap = HashMap<(u64, u64), Arc<AtomicU64>>;
```

Comment:
I guess using one map may cause too many contentions and reduce performance, maybe we should use arrays of map like Vec<Arc<RwLock<TsMap>> and hash the region id to a map at first.

Reply:
Ok. I'm getting into some problems about TiDB's code now... and after that, I'll try more different implementations here.